go/vt/proto/vtgate/vstreamer/vstreamer.go

```diff
@@ -0,0 +1,389 @@
+	ErrorOnUnknownFields bool
+
+	// this is the function that will be called to flush the rows to the handler. This is called when the
+	// minFlushDuration has passed, or the maxRowsPerFlush has been exceeded.
+	FlushFn FlushFunc
+
+	// TODO: translate this to *sql.Tx so we can pass it to the flush function
+	FlushInTx bool
+
+	// stats about flushes for this table
+	flushCount      int
+	flushedRowCount int
+	// the time this table had rows flushed, which might be different from the stream flush time
+	lastFlushedAt time.Time
+
+	// this is the data type for this table, which is used to scan the results. Regardless whether a pointer
+	// or value is supplied, the return value will always be []*DataType.
+	DataType       any
+	underlyingType reflect.Type
+
+	// this stores the current batch of rows for this table, which caches the results until the next flush.
+	// This is always a []*DataType. A nil value represents a row delete...
+	// TODO(not right...)
+	currentBatch reflect.Value
+
+	// this is the mapping of fields to the query results, which is used to scan the results. This is done
+	// on a per-shard basis, because while unlikely, it's possible that the same table in different shards
+	// could have different schemas.
+	shards map[string]shardConfig
+}
+
+// shardConfig is the per-shard configuration for a table, which is used to scan the results
+type shardConfig struct {
+	fieldMap map[string]fieldMapping
+	fields   []*querypb.Field
+}
+
+// fieldMapping caches the mapping of table fields to struct fields, to reduce reflection overhead. This is
+// configured once per shard, and used for all rows in that shard, unless a DDL event changes the schema,
+// in which case the mapping is updated.
+type fieldMapping struct {
+	rowIndex    int
+	structIndex []int
+	structType  any
+	kind        reflect.Kind
+	isPointer   bool
+}
+
+// New initializes a new VStreamer, which is used to stream binlog events from Vitess.
+func New(ctx context.Context, conn *vtgateconn.VTGateConn, tables []TableConfig, opts ...Option) (*VStreamer, error) {
+	// validate required parameters
+	if conn == nil {
+		return nil, fmt.Errorf("vstreamer: conn is required")
+	}
+
+	// initialize the VStreamer, with options and settings to be set later
+	v := &VStreamer{
+		conn:             conn,
+		session:          conn.Session("", nil),
+		tables:           make(map[string]*TableConfig),
+		minFlushDuration: 500 * time.Millisecond,
+	}
+
+	// load all shards, so we can validate settings before starting. It's not technically necessary to do this here,
+	// but it's more user-friendly to fail early if there is misconfiguration. This needs to be done before running
+	// the options, so that the shards are available for validation.
+	var err error
+	// v.shardsByKeyspace, err = getShardsByKeyspace(ctx, v.session)
+	// if err != nil {
+	// 	return nil, err
+	// }
+
+	err = v.initTables(tables)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
I think there's a consolidation issue to deal with here. Someone could easily add a new table in code and redeploy, expecting it to catch up, but it wouldn't by default. To support that, I think we'd have to store a list of tables in state, and if a new table is added, catch it up to the same vgtid, then do a cutover to be in the same stream. That'd be a much better user experience, but a decent amount of added complexity.